### ISSUE TITLE
database: implement fixed id for items surviving item renaming

### ISSUE THREAD

**Issue by bmxp** [smarthomeNG/plugins#373]
Still open:
 
* When changing item names there will be a new id which is often not really wanted. 
  --> provide a fixed database id for items

Moved to core:
 
* https://knx-user-forum.de/forum/supportforen/smarthome-py/1273010-verbrauch-gas-etc-ungewolltes-item-trigger-bei-neustart -> must be resolved in core, creating there.

Implemented (closed):
 
* https://knx-user-forum.de/forum/supportforen/smarthome-py/1278859-update-von-gleichbleibenden-messwerten-in-die-datenbank -> **implemented**
* https://github.com/smarthomeNG/smarthome/issues/165#issuecomment-280458728 -> **implemented in database plugin, needs plugins to natively support this, possible individually per item now**

New attributes: 
* ``database_write: on_update`` to enable writing of a recurrent value without a change -> behaviour **has been implemented** in v1.7.2 through the standard attribute **enforce_change**
* ``database_maxage: <days>`` maximum age of entries within a database for an item -> **has alreedy been implemented**

New feature wishes:
* compacting data -->  discuss several strategies -> **has already been implemented in develop-post_1.12.2**
* display orphaned items in database and offer to merge them into another item log -> **has been implemented**

Webinterface:
* select a datetime range for csv export -> **implemented**
* optionally select local time or UTC for csv export -> **implemented**
* format datetime as such instead of time since epoch in milliseconds
  like ``2021:11:07 12:15:12.180`` instead of ``1636243430180``  -> **implemented**

**Comment by msinn** [comment]
**Erledigt:**
> Update von gleichbleibenden Messwerten in die Datenbank

has been implemented in SmartHomeNG v1.7.2 -> new item Attribute **enforce_change**

> smarthomeNG/smarthome#165 (comment)

Issue has been closed on July, 9th 2017

> New attributes

``database_write: on_update`` Das ist über das neue Standard-Attribut ``enforce_chane`` abgedeckt
``database_maxage``Ist implementiert

**Offen**:
> https://knx-user-forum.de/forum/supportforen/smarthome-py/1273010-verbrauch-gas-etc-ungewolltes-item-trigger-bei-neustart

Ist ein Thema, aber vielleicht keines für das database Plugin. Das triggern betrifft ja nicht nur das Schreiben in die Datenbank, sondern auch die Neuberechnung anderer Items und das Triggern von Logiken

> New feature wishes:
> - compacting data
>   discuss several strategies

Das Verdichten der Daten ist problematisch, da bisher jeder mit dem ich darüber diskutierte andere Wünsche hatte und diese Wünsche sich häufig gegenseitis ausschlossen.

**Comment by Tom-Bom-badil** [comment]
Für mich ist in diesem Zusammenhang immer noch das uralte Thema "kaputte Visu-Linien" offen, siehe [hier](https://github.com/smarthomeNG/smarthome/issues/165#issuecomment-280458728). Von mir aus auch gern optional pro Item einstellbar.

Meines Wissens kann das Database-Modul immer noch keine komplette Zeitverläufe inkl. Ausfälle abbilden.

Als Workaround schleppe ich mittlerweile für alle wichtigen (numerischen/boolschen) Items ein 'alive'-Item (0/1) mit, mit dem ich den Item-Wert multipliziere. Dadurch erhalte ich den Effekt, dass ich bei Ausfall eines Gerät oder Sensor die gewünschte "0" in der Datenbank zu stehen habe, und kann auch 'sonderbare' Werte beim Initialisieren von shNG unterdrücken, und sehe sogar, wann das Gerät ausfiel und wieder 'hochkam' (z.B. ESP-Reboot nach WLAN-Verbindungsfehler).

Das hab ich damals schon ergebnislos mit Markus und einigen anderen diskutiert. Markus' Ansicht war damals, "0" könnte ja auch ein gültiger Meßwert sein und dürfe deshalb auf keinen Fall verwendet werden (nach dem Motto: Wenn nichts gemessen wird, darf auch nichts in der Datenbank stehen). Das mag streng logisch richtig sein, hilft mir aber nicht, vernünftige Plots zu erzeugen. Außerdem möchte ich gern selbst festlegen können, wie sich ein Item bei Ausfall verhalten soll.

/tom

**Comment by bmxp** [comment]
@Tom-Bom-badil löst bei Dir ein ``database: init`` das Problem mit den Nullwerten?

**Comment by Tom-Bom-badil** [comment]
Nein, die Items stehen bereits auf `database: init` - für den Plot im oben konkret verlinkten Beispiel siehe [hier](https://github.com/Tom-Bom-badil/helios/blob/a24ca82d941f7fe0e6ed8529205bba39854e25d4/files/helios.yaml#L172).

init ist aber auch nicht die Lösung für das genannte Problem - denn das Ziel ist es, Ausfallzeiten zu visualisieren. Und nicht, den Plot einfach vom letzten bekannten Zustand des Messwertes beim nächsten shNG-Reboot bzw. dem nächsten Wiedereinschalten der Gerätes als gerade Linie weiterzuziehen. Das suggeriert dem Anwender, alles wäre in Ordnung gewesen, was ja nicht stimmt.

/tom

**Comment by Morg42** [comment]
Ich verstehe das Problem nicht ganz. Wenn das Gerät offline ist, woher kommen dann die geloggten Werte? 

Sollte da nicht eher das jeweilige Plugin nicht updaten bzw. None liefern? Oder hast du einen anderen konkreten Fall?

**Comment by Tom-Bom-badil** [comment]
Nein, da dann *alle* Plugins angepasst werden müssten, statt einer einzigen Anpassung im database Plugin. Es ist ein grundsätzliches Problem.

Soweit ich am Rande mitbekommen habe, wurden ohnehin Teile dieser mehr als 10 Jahre alten Diskussion mittlerweile umgesetzt (es gibt dazu hier und drüben im KNXUF mehrere Threads, nicht nur diesen über 3 Jahre alten).

Ist aber auch egal - das System, in das ich gerade migriere, kennt die Stati 'unknown' und 'unavailable', kann diese in die Datenbank schreiben und lässt in Plots entsprechende Lücken. In meinen Augen eine sehr gute Umsetzung - statt zwischen dem letzten bekannten/gültigen Wert und dem ersten ersten neuen Wert einfach eine Linie quer über den Plot zu ziehen, ist dort einfach korrekterweise eine Lücke im Plot; auch bei kumulierten Daten, Durchschnittswerten usw wird das berücksichtigt.

Hier mal ein Bild, das beschreibt, was ich meine: ESP-Ausfall am 3. Dezember bei ca 6 Grad Celsius, repariert am 7. Dezember bei ca 10 Grad Celsius. Die lange schräge Linie zwischen diesen beiden Werten dürfte es gar nicht geben, denn da wurde nicht gemessen und somit das Temperatur-Item auch nicht aktualisiert. Cycle-Intervall der Datenquelle, last_update des Items usw waren die ganze Zeit während des Ausfalls bekannt - man könnte das also gut auswerten (wenn man denn wollte).

![image](https://github.com/user-attachments/assets/a374d776-bb99-4b79-bac8-a695a6242138)


Aber wie schon geschrieben - egal ...

/tom

**Comment by Morg42** [comment]
Okay, danke für die Erklärung.

Das wäre aus meiner Sicht eine sinnvolle, aber wahrscheinlich erhebliche Änderung am database-Plugin... aber eine Überlegung wert.

**Comment by Morg42** [comment]
Der Vollständigkeit halber - 

die Datenbank speichert zu einem Wert die Dauer der Gültigkeit. Solange zwei Stunden lang jede Minute eine 1 kommt, wird nur die Dauer der 1 verlängert, aber keine zweite 1 gespeichert.
Wenn dann eine Weile keine 1 kommt, wird auch kein Wert gespeichert und keine Dauer verlängert, da kein Item-Update das Db-Update triggert.

Insofern wären die in der Db vorhandenen Daten korrekt abbildbar, wenn beim Auslesen die Dauer "irgendwie" reproduziert wird (festes Intervall, - derzeit nicht - gespeicherte Updatezeiten, mindestens nochmal derselbe Wert am Ende der Gültigkeit (da wurde ja geschrieben, also gab es da noch ein Update)).
Das passiert derzeit nicht, insofern sind die "exportierten" Daten nicht so genau, wie sie sein könnten. Bei mir sind es derzeit z.B. PV-Produktion oder Batterieentnahme über Nacht, sobald die Batterie leer ist.

**Comment by Tom-Bom-badil** [comment]
Ich mag mich da jetzt nicht mehr reindenken, da ich im Moment jede Menge andere Baustellen habe. Es wurden aber schon Lösungsansätze recht detailliert diskutiert, u.a. [hier](https://github.com/smarthomeNG/smarthome/issues/165#issuecomment-280458728).

Ist aber alles im Sande verlaufen, weil der Anwender, diese große Wundertüte, sich über ein zusätzliches Item wundern könnte (wenn ich das richtig in Erinnerung habe) ...

/tom

**Comment by Morg42** [comment]
lib.db und plugins/database ermöglichen es, Daten als "no data" zu kennzeichnen, diese werden bei Auswertungen und Plots dann rausgerechnet.

Dies muss allerdings (wie von Tom-Bom-badil befürchtet) von allen Plugins implementiert werden. Ich mache mich demnächst erstmal an die Gateway-Plugins ran.

Damit man nicht darauf warten muss, gibt es das database-Item-Attribut `database_invalid_after`. Damit kann **für jedes einzelne Item** eine Zeitspanne definiert werden. Wenn diese Zeitspanne abläuft, ohne dass das Item eine Änderung oder ein Update bekommt, wird es automatisch auf "no data" gesetzt. Der nächste Wert wird wieder normal geschrieben.
Dies ist natürlich nur sinnvoll, wenn die Werte auch regelmäßig kommen. Beispielsweise eine Heizungsanbindung, die alle 30 Sekunden Werte liefert - nach (beispielsweise) 90 Sekunden könnte das Item auf "no data" gesetzt werden, weil die erwarteten Updates nicht gekommen sind. 

Details sind in der Doku zum database-Plugin enthalten.

Damit sind aus diesem Thema alle Features implementiert, mit zwei Ausnahmen:

1. Item-Trigger beim Neustart (kein database-Problem, sondern core)
2. Fixed database id for items

**Comment by Morg42** [comment]
Sofern zum Umbenennen der Items nicht die yaml-Konfiguration direkt bearbeitet wird, sondern das AdminUI verwendet wird, werden die Datenbank-IDs automatisch angepasst, und es ist keine weitere Arbeit notwendig.

Ob sich die Implementierung einer persistenten Datenbank-ID (samt Management-Methoden) lohnt, diskutieren wir noch.
